### DISCUSSION TITLE
deleteByQuery, updateByQuery

### DISCUSSION THREAD

**Discussion by claytongulick** [directus/directus#17489] [Feature Requests]
### Summary

It feels strange, and like a potential source of timing related bugs to have to do two roundtrips to the server to do the equivalent of a deleteByQuery from the SDK.

Currently what we need to do is formulate a readByQuery to return the IDs to delete, and then pass those IDs to a deleteMany call. In a medium to high volume transaction system many records could be missed during the round trip time gathering the IDs and then calling deleteMany, leaving the system in an unpredictable state.

Looking at the source of the items service, I see deleteByQuery is also implemented this way, it retrieves the IDs to delete and then passes them to deleteMany. It looks like there's a potential bug with this implementation as well, because the call to getKeysByQuery and deleteMany don't participate in the same transaction, causing a potential timing related inconsistent state.

The purpose of this discussion is to think through:
1) Should there by a deleteByQuery & updateByQuery in the SDK?
2) Is there a more robust implementation of deleteByQuery & updateByQuery in the items service? Is the current approach of gathering IDs to hand to deleteMany the best approach, or should the AST support creating real SQL DML statements for deletes/updates?

Thanks!

### Motivation

The motivation is to expand the functionality of delete and update to be consistent with the experience of reads, and also prevent a class of timing related bugs that can arise from the current implementation.

### Detailed Design

I believe the AST would need to be expanded to support the parsing of delete and update queries to create DELETE/UPDATE DML statements with the same degree of robust permissions and logging that are supported by read operations.

The implementation would necessarily involve continuing the approach of getKeysByQuery in order to properly execute activity logging and hooks, but it would properly disconnect those functions from the delete operation itself. getKeyByQuery would need to be enhanced to accept an optional knex trx to participate in.

### Must Have:
- A transaction aware deleteByQuery implementation in the items service that prevents timing related data bugs.
- An implementation in the SDK to invoke the service

### Could Have:
- The same thing for updates?

### Drawbacks

This is an area the Directus team can give the best feedback on. If it were simple, I'm sure it would already be implemented.

### Alternatives

An alternative would be to just add the deleteByQuery and updateByQuery functions to the SDK that call the existing implementation in the items service.

I do think it's important to address the transaction issue though, because unless I'm misreading or misunderstanding the code there's a potential for data inconsistency with timing based on readKeysByQuery and deleteMany/updateMany not participating in the same transaction.

### Adoption Strategy

This shouldn't be a breaking change, it's adding additional functionality.

**Comment by rijkvanzanten** [comment]
> Should there by a deleteByQuery & updateByQuery in the SDK?

I think so! The REST API already supports *byQuery, so it makes sense to mirror that in the SDK

> Is there a more robust implementation of deleteByQuery & updateByQuery in the items service?

I believe you already answered your own question 🙂  The current implementation is.. the implementation. There's no hidden one elsewhere if that's what you mean 🤔 

> Is the current approach of gathering IDs to hand to deleteMany the best approach, or should the AST support creating real SQL DML statements for deletes/updates?

It's not the _best_ approach, but it was a reasonably decent approach for a lot of use cases, which was possible to create in a reasonable effort. Only queries (eg `read`) are currently relying on the full AST based system[^1] which in turn means we were bound to using that to first extract which items to update/delete, after which we could use the "regular" update mechanisms to execute the mutation. That being said, we are currently in the middle of designing out what this AST structure would look like for mutations, at which point it also becomes possible to do the _best_ solution here (which would be to generate out a proper `DELETE FROM x WHERE <full filters>` SQL statement 🙂 

[^1]: Read requests from rest/graphql/ws are converted to an AST early, which is then passed through the services pipeline to be validated and executed as SQL. Writes however are passed in as an object of changed columns rather than a full AST, so don't have the same filtering capabilities baked in at the moment.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
